src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -22057,5 +22321,910 @@ public interface IS2
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "[outer] = inner").WithArguments("inner").WithLocation(29, 29)
                 );
         }
+
+        [Fact]
+        public void NullCheck_01()
+        {
+            var src = @"
+public class Helper
+{
+    public static bool Test1<T>(T value)
+        where T : allows ref struct
+    {
+        return value == null;
+    }
+
+    public static bool Test2<T>(T value)
+        where T : allows ref struct
+    {
+        return null == value;
+    }
+
+    public static bool Test3<T>(T value)
+        where T : allows ref struct
+    {
+        return value != null;
+    }
+
+    public static bool Test4<T>(T value)
+        where T : allows ref struct
+    {
+        return null != value;
+    }
+}
+
+ref struct S
+{
+}
+
+class Program
+{
+    static void Main()
+    {
+        System.Console.Write(Helper.Test1<S>(new S()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test2<S>(new S()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test3<S>(new S()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test4<S>(new S()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test1<Program>(null));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test2<Program>(null));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test3<Program>(null));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test4<Program>(null));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test1<Program>(new Program()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test2<Program>(new Program()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test3<Program>(new Program()));
+        System.Console.Write(' ');
+        System.Console.Write(Helper.Test4<Program>(new Program()));
+    }
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "False False True True True True False False False False True True" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+
+            verifier.VerifyIL("Helper.Test1<T>(T)",
+@"
+{
+  // Code size       12 (0xc)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""T""
+  IL_0006:  brtrue.s   IL_000a
```

Comment:
Why don't we emit branchless code like `box; ldnull; ceq` instead?

Reply:
Why don't we emit branchless code like ```box; ldnull; ceq``` instead?

Because that IL sequence is illegal for type parameters that allow ref struct. Boxing is not supported for them in general, but an exception is made for `box, brtue/brfalse` sequence.